diff --git a/drivers/usb/gadget/android.c b/drivers/usb/gadget/android.c
index f79512a35c6f..06f13dbda80c 100644
--- a/drivers/usb/gadget/android.c
+++ b/drivers/usb/gadget/android.c
@@ -53,6 +53,7 @@
 #include "f_accessory.c"
 #define USB_ETH_RNDIS y
 #include "f_rndis.c"
+#include "f_ecm.c"
 #include "rndis.c"
 #include "u_ether.c"
 #include "f_dm.c"
@@ -523,6 +524,62 @@ static struct android_usb_function rndis_function = {
 };
 
 
+struct ecm_function_config {
+	u8 ethaddr[ETH_ALEN];
+};
+
+static int ecm_function_init(struct android_usb_function *f,
+			     struct usb_composite_dev *cdev)
+{
+	struct ecm_function_config *config;
+
+	config = kzalloc(sizeof(*config), GFP_KERNEL);
+	if (!config)
+		return -ENOMEM;
+
+	config->ethaddr[0] = 0x02;
+	config->ethaddr[1] = 0x12;
+	config->ethaddr[2] = 0x34;
+	config->ethaddr[3] = 0x56;
+	config->ethaddr[4] = 0x78;
+	config->ethaddr[5] = 0x9a;
+	f->config = config;
+	return 0;
+}
+
+static void ecm_function_cleanup(struct android_usb_function *f)
+{
+	kfree(f->config);
+	f->config = NULL;
+}
+
+static int ecm_function_bind_config(struct android_usb_function *f,
+				    struct usb_configuration *c)
+{
+	struct ecm_function_config *config = f->config;
+	int ret;
+
+	ret = gether_setup_name(c->cdev->gadget, config->ethaddr, "usb");
+	if (ret)
+		return ret;
+	return ecm_bind_config(c, config->ethaddr);
+}
+
+static void ecm_function_unbind_config(struct android_usb_function *f,
+				       struct usb_configuration *c)
+{
+	gether_cleanup();
+}
+
+static struct android_usb_function ecm_function = {
+	.name		= "ecm",
+	.init		= ecm_function_init,
+	.cleanup	= ecm_function_cleanup,
+	.bind_config	= ecm_function_bind_config,
+	.unbind_config	= ecm_function_unbind_config,
+};
+
+
 struct mass_storage_function_config {
 	struct fsg_config fsg;
 	struct fsg_common *common;
@@ -663,6 +719,7 @@ static struct android_usb_function *supported_functions[] = {
 	&mtp_function,
 	&ptp_function,
 	&rndis_function,
+	&ecm_function,
 	&mass_storage_function,
 	&accessory_function,
 	&dm_function,
